src/libmongoc/tests/unified/entity-map.c

```diff
@@ -479,6 +480,11 @@ apm_command_callback (command_callback_funcs_t funcs, const char *type, const vo
    BSON_ASSERT (funcs.get_context);
    entity_t *const entity = (entity_t *) funcs.get_context (apm_command);
 
+   // Check if callbacks are disabled
+   if (bson_atomic_int_fetch (&entity->callbacks_disabled, bson_memory_order_relaxed)) {
```

Comment:
On further read, I am concerned about accessing shared data within callbacks.

I do not think it is a problem yet within this PR, since the only monitored events are command events. But I expect adding SDAM events may result in shared access:

With a client pool, callbacks may be called from separate threads. Example:
- A server monitor thread calls: `_server_monitor_thread` => `_server_monitor_check_server` => `_server_monitor_heartbeat_started` (calls heartbeat started callback)
- Meanwhile, the main thread calls: `mongoc_client_command_simple` => `_mongoc_client_command_with_stream` => `mongoc_cluster_run_command_monitored` (calls command started callback)

This may result in new inconsistent test failures (e.g. missed events due to a race).

Identifying each case of shared access and locking may be a significant effort, and difficult to get right. Since using a client pool is only needed for the spec test `serverMonitoringMode.yml`, I would rather implement the functionality of that spec test separately. Example:

```c
// `smm_t` is a test fixture for serverMonitoringMode tests.
#define MAX_EVENTS 3
typedef struct {
   bson_mutex_t lock;
   bool awaited[MAX_EVENTS];
   size_t awaited_len;
   mongoc_client_pool_t *pool;
} smm_t;

static void
heartbeat_started (const mongoc_apm_server_heartbeat_started_t *event)
{
   smm_t *t = mongoc_apm_server_heartbeat_started_get_context (event);
   bson_mutex_lock (&t->lock);
   // Store the most recent awaited value (if room).
   if (t->awaited_len < MAX_EVENTS) {
      bool awaited = mongoc_apm_server_heartbeat_started_get_awaited (event);
      t->awaited[t->awaited_len] = awaited;
      t->awaited_len++;
   }
   bson_mutex_unlock (&t->lock);
}

static smm_t *
smm_new (const char *mode)
{
   smm_t *t = bson_malloc0 (sizeof (smm_t));
   bson_mutex_init (&t->lock);

   // Create client pool.
   {
      mongoc_uri_t *uri = test_framework_get_uri ();
      mongoc_uri_set_option_as_int32 (uri, MONGOC_URI_HEARTBEATFREQUENCYMS, 500); // To speed up test.
      mongoc_uri_set_option_as_utf8 (uri, "serverMonitoringMode", mode);
      t->pool = mongoc_client_pool_new (uri);
      test_framework_set_pool_ssl_opts (t->pool);
      mongoc_uri_destroy (uri);
   }

   // Capture heartbeat started events.
   {
      mongoc_apm_callbacks_t *cbs = mongoc_apm_callbacks_new ();
      mongoc_apm_set_server_heartbeat_started_cb (cbs, heartbeat_started);
      mongoc_client_pool_set_apm_callbacks (t->pool, cbs, t);
      mongoc_apm_callbacks_destroy (cbs);
   }

   // Pop and push a client to start monitoring.
   {
      mongoc_client_t *client = mongoc_client_pool_pop (t->pool);
      mongoc_client_pool_push (t->pool, client);
   }

   return t;
}

// `smm_wait` waits for `count` heartbeat started events.
static bool
smm_wait (smm_t *t, size_t count)
{
   int64_t started = bson_get_monotonic_time ();
   while (true) {
      bson_mutex_lock (&t->lock);
      if (t->awaited_len >= count) {
         bson_mutex_unlock (&t->lock);
         return true;
      }
      bson_mutex_unlock (&t->lock);

      int64_t now = bson_get_monotonic_time ();
      if (now - started > 10 * 1000 * 1000) {
         break;
      }
      _mongoc_usleep (500 * 1000); // Sleep for 500ms.
   }
   return false;
}

static void
smm_destroy (smm_t *t)
{
   if (!t) {
      return;
   }
   mongoc_client_pool_destroy (t->pool);
   bson_mutex_destroy (&t->lock);
   bson_free (t);
}

// `test_serverMonitoringMode` implements spec tests from serverMonitoringMode.yml.
// The spec test needs a client pool for stream monitoring. The unified test runner only uses single-threaded clients.
static void
test_serverMonitoringMode (void *unused)
{
   BSON_UNUSED (unused);

   if (test_framework_is_replset ()) {
      printf ("Test is skipped. SDAM events are non-deterministic when monitoring multiple servers.");
      return;
   }

   if (test_framework_get_server_version () >= test_framework_str_to_version ("4.4.0")) {
      printf ("'connect with serverMonitoringMode=auto >=4.4' ... begin\n");
      smm_t *t = smm_new ("auto");
      ASSERT (smm_wait (t, 2));
      // Check events.
      {
         bson_mutex_lock (&t->lock);
         ASSERT_CMPSIZE_T (t->awaited_len, >=, 2);
         ASSERT (!t->awaited[0]); // Polls.
         ASSERT (t->awaited[1]); // Streams.
         bson_mutex_unlock (&t->lock);
      }
      smm_destroy (t);
      printf ("'connect with serverMonitoringMode=auto >=4.4' ... end\n");
   }

   // TODO: other tests.
}
```